NebulaWorld/Player/GizmoManager.cs

```diff
@@ -0,0 +1,270 @@
+﻿#region
+
+using System;
+using System.Text;
+using NebulaWorld.Chat.ChatLinks;
+using NebulaWorld.MonoBehaviours.Local.Chat;
+using UnityEngine;
+#pragma warning disable IDE1006 // Naming Styles
+
+#endregion
+
+namespace NebulaWorld.Player;
+
+public class GizmoManager : IDisposable
+{
+    public ushort ObservingPlayerId { get; set; }
+    public int ObservingPlanetId { get; set; }
+    public Vector3 ObservingPos { get; set; }
+
+    private ushort indicatorPlayerId;
+    private int indicatorPlanetId;
+    private Vector3 indicatorPos;
+    private LineGizmo naviIndicatorGizmo;
+    private LineGizmo naviIndicatorGizmoStarmap;
+    private CircleGizmo targetGizmo0;
+    private CircleGizmo targetGizmo1;
+
+    public void Dispose()
+    {
+        naviIndicatorGizmo = null;
+        naviIndicatorGizmoStarmap = null;
+        targetGizmo0 = null;
+        targetGizmo1 = null;
+        GC.SuppressFinalize(this);
+    }
+
+    public void SetIndicatorPlayerId(ushort playerId)
+    {
+        CloseTargetGizmos();
+        if (indicatorPlayerId == playerId)
+        {
+            indicatorPlayerId = 0;
+            return;
+        }
+        indicatorPlayerId = playerId;
+
+        indicatorPlanetId = 0;
+        GameMain.mainPlayer.navigation.indicatorAstroId = 0;
+        GameMain.mainPlayer.navigation.indicatorMsgId = 0;
+        GameMain.mainPlayer.navigation.indicatorEnemyId = 0;
+    }
+
+    public void SetIndicatorPing(int planetId, Vector3 pos)
+    {
+        CloseTargetGizmos();
+        if (indicatorPlanetId == planetId)
+        {
+            indicatorPlanetId = 0;
+            return;
+        }
+        indicatorPlanetId = planetId;
+        indicatorPos = pos;
+
+        indicatorPlayerId = 0;
+        GameMain.mainPlayer.navigation.indicatorAstroId = 0;
+        GameMain.mainPlayer.navigation.indicatorMsgId = 0;
+        GameMain.mainPlayer.navigation.indicatorEnemyId = 0;
+    }
+
+    public void OnUpdate()
+    {
+        if (VFInput.alt && VFInput.control && Input.GetMouseButtonDown(0)) GetMapPing();
+        UpdateIndicator();
+    }
+
+    private static void GetMapPing()
+    {
+        // Modify from UIGlobemap.TeleportLogic
+        if (Camera.main == null || GameMain.localPlanet == null) return;
```

Comment:
```suggestion
		// Modify from UIGlobemap.TeleportLogic
        var mainCam = Camera.main;
        if (mainCam == null || GameMain.localPlanet == null) return;
        if (!Physics.Raycast(mainCam.ScreenPointToRay(Input.mousePosition), out var hitInfo, 800f, 8720, QueryTriggerInteraction.Collide)) return;
```

The unity analyzer says grabbing Camera.main is expensive, so while it probably doesn't matter for this method, let's store it in a variable for quicker access